### ISSUE TITLE
EventedList does not support assignment to an item by lookup

### ISSUE THREAD

**Issue by andy-sweet** [napari/napari#3058]
## 🐛 Bug

`EventedList` does not support assignment to an item when using a custom `lookup`, though does support access to that element through the same lookup. For example, `TransformChain` is an `EventedList` that has `lookup` based on the name of the transform. As described in #3045 `transform['physical2world'].affine_matrix = ...` works, but `transform['physical2world'] = ...` does not.

## To Reproduce

Here's a simple test case that currently fails at HEAD.

```python
def test_transform_chain_assignment_by_name():
    scale_2 = Affine(scale=(2, 2), name='scale')
    scale_3 = Affine(scale=(3, 3), name='scale')
    chain = TransformChain([scale_2])

    chain['scale'] = scale_3

    assert chain['scale'] == scale_3
```

## Expected behavior

I would expect both `__getitem__` and `__setitem__` to support indexing by `lookup` or neither.

## Environment

napari: 0.4.9rc1.dev70+g5fe37b47.d20210709
Platform: macOS-11.3.1-x86_64-i386-64bit
System: MacOS 11.3.1
Python: 3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:15) [Clang 11.1.0 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.19.5
SciPy: 1.6.3
Dask: 2021.04.1
VisPy: 0.6.6

OpenGL:
- GL version: 2.1 INTEL-16.2.16
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 2560x1440, scale 1.0

Plugins:
- brainglobe-io: 0.0.2
- brainreg-standard: 0.0.2
- cellfinder: 0.0.10
- console: 0.0.3
- napari-demo: 0.0.2
- napari-properties-plotter: 0.1.0
- napari-properties-viewer: 0.0.2
- napari-skimage-regionprops: 0.1.0
- ndtiffs: 0.1.1
- scikit-image: 0.4.9rc1.dev70+g5fe37b47.d20210709
- svg: 0.1.5

**Comment by tlambert03** [comment]
to be honest, I've never liked that "feature".  I feel like that's a mutable mapping... not a mutable sequence.  I do see that the asymmetry is probably stranger... but I'd rather see us _remove_ the ability to get/set on a key rather than double down on it.
I know we'd need to "grandfather in" the layerlist... but there too, the fact that the name key has to be unique makes for a lot of unfortunate one-off special casing that would have been avoided if it was just an ordered dict.  
Can you refactor `TransformChain` to either not use the keys for lookup, or perhaps use an evented dict?  Or, even better... if you _know_ you have 3 or 4 specific named transforms that are always part of the model, make a Pydantic model with each of them as a field?
```py
class LayerTransforms(EventedModel, Transform):
    tile2data: Transform
    data2physical: Transform
    physical2world: Transform
    world2grid: Transform
```

**Comment by tlambert03** [comment]
i think this discussion is relevant too here: https://github.com/napari/napari/pull/2793#discussion_r646144451

**Comment by andy-sweet** [comment]
Removing the behavior is good with me, as I agree the custom key mapping for a list is unexpected.

Like you mentioned `LayerList` also shares this behavior, so the test below fails too, but I don't think we depend on that behavior anywhere in our implementation and it probably isn't a high priority for users either (though accessing by layer names is!).

```python
def test_assignment_by_name():
    layers = LayerList()
    image1 = Image(np.random.random((10, 10)), name='image')
    layers.append(image1)

    image2 = Image(np.random.random((10, 10)), name='image')
    layers['image'] = image2

    assert layers['image'] == image2
```

Agreed that named fields in `TransformChain` is a straightforward solution to avoid this buggy behavior for that class in particular, but I think it's worth keeping this open to resolve the asymmetry in `EventedList` and to work out what to do with `LayersList`.

If we wanted to use an `EventedDict` (for either `TransformChain` or `LayersList`) can we rely on the insertion order being maintained? I think this might reduce to whether we only support Python 3.7 and above, but unsure.

**Comment by tlambert03** [comment]
> Like you mentioned LayerList also shares this behavior, so the test below fails too

yep, and given the hurdles we already jump through to make `layers['name']` work (i.e renaming the layer if necessary) I think `layers['name'] = Image(... name='other name')` probably _should_ continue to be an error...  Basically, it's just a list with (unfortunately) a custom accessor... but not a dict.  

Happy to hear what others have to say here; but I doubt anyone wants to remove the `layers['name']` accessor any time soon (I'd be in support of it though!)... and I'd definitely try to sway us away from supporting "appending" a layer with `__setitem__`.

> If we wanted to use an EventedDict (for either TransformChain or LayersList) can we rely on the insertion order being maintained?

yep, the [underlying data store](https://github.com/napari/napari/blob/3c8c68c7af52889e72deb6816ba0840221709d83/napari/utils/events/containers/_dict.py#L29) is just a regular python dict, which will be ordered for all versions of python we support.

That said... i'd strongly prefer that we not use EventedDict when the keys are known ahead of time.  Pydantic models are definitely preferred there.

**Comment by sofroniewn** [comment]
> Removing the behavior is good with me, as I agree the custom key mapping for a list is unexpected.

fine with me too remove in general/ and in transform chain - but

> Happy to hear what others have to say here; but I doubt anyone wants to remove the layers['name'] accessor any time soon (I'd be in support of it though!)... and I'd definitely try to sway us away from supporting "appending" a layer with __setitem__.

it might be nice to keep this at least for a while longer - I do use it quite a lot, and think others do too, but I get the concerns here too ...

**Comment by willingc** [comment]
I'm going to relabel this as refactor instead of bug.
